server/schedule/checker/rule_checker.go

```diff
@@ -222,11 +223,26 @@ func (c *RuleChecker) replaceUnexpectRulePeer(region *core.RegionInfo, rf *place
 		}
 	}
 
+	var witness *metapb.Peer
+	for _, witness = range region.GetWitnesses() {
+		if witness.StoreId != peer.StoreId {
+			break
+		}
+	}
+
 	createOp := func() (*operator.Operator, error) {
 		if newLeader != nil && newLeader.GetId() != peer.GetId() {
 			return operator.CreateReplaceLeaderPeerOperator("replace-rule-"+status+"-leader-peer", c.cluster, region, operator.OpReplica, peer.StoreId, newPeer, newLeader)
 		}
-		return operator.CreateMovePeerOperator("replace-rule-"+status+"-peer", c.cluster, region, operator.OpReplica, peer.StoreId, newPeer)
+		/* TODO: enable this after the pr  balance witness merged
+		if newPeer.IsWitness {
+			return operator.CreateMoveWitnessOperator("replace-rule-"+status+"-witness-peer", c.cluster, region, peer.StoreId, newPeer.StoreId, operator.OpWitness)
+		}
+		*/
+		if witness == nil {
```

Comment:
What if a peer is both witness and leader?

Reply:
TiKV will forbid change leader to witness, but when the leader is down and the witness has more logs, the witness becomes the leader, then proposes a transfer leader command. So in this process, pd may know that the peer may have dual identities, right?

I plan to add a choose condition in:

```
for _, witness = range region.GetWitnesses() {
    if witness.StoreId != peer.StoreId && witness.Id != leaderId {
        break;
   }
}
```

Does this make sense?